### ISSUE TITLE
stripe integration

### ISSUE THREAD

**Issue by scd** [theCrag/website#1841]
Review stripe capabilities:
 - [x] api based payment gateway means we can get fully integrated credit card payments
 - [x] we don't need to see credit card details on our server
 - [x] multi - currency automatically converted to AUD
 - [x] I think payments automatically go to my bank account, which might increase the admin overhead. But I do have an accounting system that can automatically tag transactions.
 - [ ] Not sure if we can tag Australian transactions differently (Australian transactions require us to collect GST, whereas OS transactions do not).
 - [x] We can create once off payments, create a customer so that they only need their credit card details once and recurring payments based on a plan.
 - [x] Stripe has an awsome test mode where you use a test api key and the charges are not really made but otherwise functions normally

Setup
 - [x] Create a stripe account
 - [x] Find a perl module
 - [x] test script to demonstrate perl module and stripe account works
 - [x] install perl module everywhere
 - [x] install api keys

Business decisions for supporter program
 - [x] Should we totally replace PayPal. I think yes - having options can be confusing as well as increasing our code maintenance load
 - [x] Do we want to set up a monthly plan for those willing to provide extra financial support. Maybe, but I don't want to do anything less than $10 per month as the transaction costs become significant. I think we have to pre-create plans as well, or create a plan on-the-fly with the amount they put in.
 - [x] Do we want to create people as customers so they only need their credit card once. 

Advertisers business rules
 - [ ] Stripe will be an enabler to redevelop advertisers so that advertising can be a self service.

Subscriber work flow notes.
 * Become a premium supporter page. Credit card form integrated on the page. If payed before show last four digits of previous credit card payment and offer a pay using previous credit card button (check expiry date?, or credit card still valid - possibly an API call ). 
 * Standard stripe flow is that when somebody submits their credit card details, a piece of javascript intervenes and sends the credit card details to stripe returning a once off token. Upon success we send the token to our server to make the charge (rather then the credit card details).
 * We should send the token to the server via our api so that any card errors can be shown on the page without having to reload the page.
 * When we make the charge on the server also create a customer. Against the our account store the following additional info - last four digits of credit card, expire date (??? - don't know if we have this), customer id.

Optional monthly plan
  * allow an optional monthly plan button for those really blown away by thecrag.

**Comment by scd** [comment]
Totally removing PayPal from the donate page - committing to Stripe only.

**Comment by scd** [comment]
I have been thinking about whether we create a customer so that our users can make subsequent payments without entering their credit card. I am thinking we do not do this, as we are only expecting one payment per year so we should keep it simple.

But I do think we should offer a monthly recurring payment of $10 or more for people who are extra enthusiastic to help. Note that this is different from creating a customer.

**Comment by scd** [comment]
Notes to self:
 - [ ] should we send email address with charge
 - [ ] before release test using live key on dev server to make sure back account is set up correctly

**Comment by scd** [comment]
@brendanheywood I have done single payment integration to stripe on our donate page. Could you please pull from repo.

I am happy to discuss the implementation. The html markup is similar to ascent logging, so it should be responsive. Feel free to make any UX improvements. At this stage I have concentrated on functionality.

Question. Is it worthwhile implementing a monthly payment option now? I think we can wait for this a little. 

Do we want an other amount or more checkboxes?

**Comment by brendanheywood** [comment]
done, error: Error, no stripe publishable key.

**Comment by scd** [comment]
now try

**Comment by scd** [comment]
@brendanheywood please pull again, I have added a couple of extra fields

**Comment by brendanheywood** [comment]
done. I'll think about ux a bit more later but some small things:

* [ ] year and month expiry should be selects

**Comment by scd** [comment]
month year is from the stripe sample code. I am wondering why they have not done this in their sample code, given that they seem best practice in everything.

**Comment by scd** [comment]
Also the card number, name, cvc and expiry date do not go to our server at all, they go directly to stripe where a token is created. The token then goes to our server. Note that for PCI compliance we cannot let any card info go to our server.

So whatever changes we make to the expiry date must be readable by the stripe js code when getting a token. Given that stripe have not offered it as part of their sample code, I recon we just run with what they have given. If it is really unacceptable then I have seen other third party code examples using select for the expiry date, so it should be ok. Brendan UXs are not really my thing so I will leave this as your final call, as long as you know the full context.

**Comment by brendanheywood** [comment]
There are heaps of little ways this could be better, I wouldn't take the example as gospel at all

* [ ] add markup to change the keyboard type and number only regex when on mobile to make typing numbers easier ```<input type=text pattern="[0-9]{13,16}">```
* [ ]  add markup hints to help chrome auto fill credit cards details from wallet http://stackoverflow.com/questions/15168261/how-does-chrome-detect-credit-card-fields
* [ ]  add markup hints to help iPhone scana CC via camera http://stackoverflow.com/questions/25163891/when-scan-credit-card-option-is-available-on-ios8-safari

http://gigliwood.com/blog/designing-the-perfect-credi.html

**Comment by brendanheywood** [comment]
It's trivial to debug what gets sent over the wire to make sure stripe gets the right stuff, and there are heaps of examples in the wild of selects with stripe like the last link example I sent above

**Comment by scd** [comment]
I made the minor changes you suggested and as it turned out the select was trivial as stripe just worked with a select field.

I have had a look at the current payment page the article is based on. It is looking cleaner than in the article. They have dumped descriptions and a few other things.

**Comment by scd** [comment]
@brendanheywood do you want me to polish the ux for the donate page. At the moment I have not done any polishing, but if you are too busy then I should do it. I think people will mostly pay using their pc, so we should make sure it looks good using bigger screen sizes.

**Comment by brendanheywood** [comment]
I had a quick play last night starting with adding the .bs-form class to the form element to turn on bootstrap styles. But I didn't get very far and kinda got bogged down investigating how hard it would be to convert to bootstrap3. so yeah go ahead and start polishing I'll chip in when I can

A few ideas:
* [ ] disable the donate button until required stuff is filled out.
* [ ] make the link to stripe open in a new window
* [ ] when you are done don't redirect back to this page, need a page some details of what it means and a big thanks - perhaps the /settings/contributions page could fullfill this role
* [ ] make sure all fields are big enough to contain typical data, email is often way too small
* [ ] the messages next to each payment amount disapear when you change current and you never get them back even when back to AUD
* [ ] the validation is a bit clunky, ie if the card is missing or wrong it should focus on that field and add the highlight it red outline (there are existing classes for this in bootstrap)
* [ ] I really like the chunks of text and think we could flesh these out more, especially with more options or an 'other' option. because they are vertical and not horizontal we have a lot more room to play with and I think it would be fine have another 1 or 2 higher options as well as other.

* [ ] one idea I've seen in a few places in the wild is an extra free text field where people can write a personal message to go with their donation. makes it more personal and also if they are prompted could give us potentially better feedback and who and why people may contribute

**Comment by scd** [comment]
Noooo not bootstrap3 just yet. I think we need it, but we need to get through a few things first.

Nice feedback.


> the validation is a bit clunky, ie if the card is missing or wrong it should focus on that field and add  > the highlight it red outline (there are existing classes for this in bootstrap)

I will see what we can do with the error messages we get back from stripe. Maybe there is a way of working out which field the error message applies to. There are only three fields where there could be a credit card error from stripe - cc#, cvc, expiry. Name is optional and will not cause the credit card to fail. We could do some extra javascript validation for some simple errors.


> I really like the chunks of text and think we could flesh these out more, especially with more options > or an 'other' option. because they are vertical and not horizontal we have a lot more room to play > with and I think it would be fine have another 1 or 2 higher options as well as other.

I liked the chunks of text because they mapped to nice distinct levels. How do you feel about launching with less rather than more just to see how it goes. Also if somebody wants to donate more than $100, then I think we should be pushing them toward a monthly payment starting at say $10. 

Maybe we have a more button which opens up other options and monthly payments.

**Comment by brendanheywood** [comment]
I'm cool with monthly but doesn't that mean we have to store the customer details for regular payments? I want this regardless anyway whether its monthly or yearly. This isn't a big or complex form so I want everything up front and not hidden behind options. I want zero barriers between caring people with their wallets ready and us.

In terms of priorities and minimum viable product, I think the 'other' is a must have, and the recurrent stuff is second priority and other levels are something we can tune easily as we go.

**Comment by scd** [comment]
> I'm cool with monthly but doesn't that mean we have to store the customer details for regular payments? 

This is where stripe is so good, they store everything. There are three levels we can engage at:
 1. Once off transaction
 2. Create a customer of ours on stripe and get stripe to remember the credit card details for later payments. We need to store the customer id for this, which is not credit card information. The customer id is only useful with our secret key.
 3. Create a plan and get somebody to subscribe to the plan. In this instance there are web hooks after each payment.

If we create a customer then it would make it easier for them to pay again in 12 months using the exisiting credit card. It adds a little complexity to the back end processing, which is probably fairly trivial to implement (maybe an additional field), however it does make the UX look like we are storing credit cards.

> This isn't a big or complex form so I want everything up front and not hidden behind options

Good to know your position here - I will get behind this.

> I think the 'other' is a must have

OK, but let's keep discussing the monthly stuff. I am really in two minds about going to the extra effort for monthly.

**Comment by brendanheywood** [comment]
I want the UX to store the credit card so it's all automated. I cannot see how anybody would sign up to a monthly supporter program without this, we'd have to email them every month and get them to fill the details in again each time. utter PITA. Any security concerns are the same as any other site and I think we can allay this with good copy,

I think I want something like "Once off | Each month", if "Once off" is selected then there is an option tick box: "[ ] Store your card securely for next time", but if the "Each month" is selected then this gets set for you and you can't untick it, or we swap it our with "You card will be securely stored using Stripe"

The other piece of this puzzle is we need to be really transparent around when the next payment will come out, probably on both this page and on the /settings/contributions page. I think stripe will automatically send receipts after each payment? If not we should, and it probably can't hurt to send a courtesy reminder that we will process a payment maybe maybe a week out. There are also other things like denied cards but I think stripe can handle all that for us too.

It should also be really easy to change the support plan, say from $50 to $100 or vice versa at any time. Ideally the exact same page just with the credit card section replaced with the summary of your stored card (last 4 digits etc), and the "Donate $50 Now" is swapped out to something else. Doing this will only change the next scheduled payment and not do anything now.

* [ ] Another tiny thing: drop the default selected month, it's defaulting to this month which has zero correlation to when a cc might expire. also add empty default options so it's clear you need to add them as they are blank

**Comment by scd** [comment]
> I want the UX to store the credit card so it's all automated. I cannot see how anybody would sign up to a monthly supporter program without this, we'd have to email them every month and get them to fill the details in again each time. utter PITA. Any security concerns are the same as any other site and I think we can allay this with good copy.

I don't think I have explained stripe properly. Firstly we cannot store the credit card  - the credit card cannot even go to our server. If either of these things happen we are not PCI compliant which is a serious breach. Stripe let's us do everything we want without ever seeing the credit card - it is awesome. The credit card goes direct to stripe servers from javascript returning a temporary token. We send the token to our server. We can charge that token. We can also set up the user as a customer on stripe, where stripe remembers the credit card, but we can never get it. We can even make subsequent charges to that credit card even though we don't have the credit card details (I think).

We can get the last four digits of the credit card though. So we could do something like '[ ] use the previous credit card ... 1234'.

This means forget any solution where we know the credit card details. However we can design the UX so that the customer only needs to enter the credit card once.

As far a monthly payments go, stripe can deal with all that. We just tell stripe that this guy has signed up for a monthly payment. 

> It should also be really easy to change the support plan, say from $50 to $100 or vice versa at any time. Ideally the exact same page just with the credit card section replaced with the summary of your stored card (last 4 digits etc), and the "Donate $50 Now" is swapped out to something else. Doing this will only change the next scheduled payment and not do anything now

We have either monthly payments or once off 12 month payments. No automated 12 month payments as it will cause way too many disputes. I have spent hours talking to somebody who has done 12 month automated renewal payments and it caused all sorts of problems and is ripping it out of his site. People don't remember what they have done.

We can still use the last 4 digits of the previous credit card to make the next once off payment. You would not swap out the credit card details, but rather have the option to use the last credit card. 

>  Another tiny thing: drop the default selected month, it's defaulting to this month which has zero correlation to when a cc might expire. also add empty default options so it's clear you need to add them as they are blank

I have been going to and fro with this one. Originally I had it blank and then looked at the good practice article (well actually their live site) to see what they did. They default to current month, which as you said has zero correlation was an enigma to me. They have obviously done it for a reason, maybe something to do with how people recognise the field. I thought the UX looked more friendly without a blank. I honestly don't care either way, I just want to settle on one way of doing it.

**Comment by brendanheywood** [comment]
I know exactly how stripe works, it's fundamentally the same as every other bank just done better and I've been through this process many times with work with multiple payment gateways. I was talking conceptually, and from my point of view and more importantly our users, whether it is stored with stripe or with us is an implementation detail. Everything I said above still stands

Stored card details can be returned including the last 4 digits: (we would look this up live, and not store it)
https://stripe.com/docs/api#retrieve_card

and you can update a card:
https://stripe.com/docs/api#update_card

> We have either monthly payments or once off 12 month payments. No automated 12 month payments as it will cause way too many disputes. 

This is a fair point. I would say it purely comes down to communications and expectations and is dependant on each service. More or less the same as what I was saying above about being really transparent so there are no surprises. Flickr stores my card and bills me every 2 years, but I manually press the pay button even though they store my card. They send me an email about a month out and then another one a week out and so on if I haven't renewed. So maybe that hybrid option is a better way where they still need to explicitly pay but we do as much as we can to make that process easy.

But yes I agree that monthly is preferable from our point of view. Maybe we need something like this:
![image](https://cloud.githubusercontent.com/assets/187449/6052062/64d52a2c-ad22-11e4-8031-70cea3f39034.png)

but then if you swap to monthly you get:
![image](https://cloud.githubusercontent.com/assets/187449/6052079/a0c21d74-ad22-11e4-9b99-bc139c29d5f7.png)

If someone is giving monthly, out of $5 38c goes to stripe which still seems very reasonable to me, so I don't see a problem with $5 being the lowest level for ongoing support. $20 a month = $240 a year which is on par with our highest donors so far so seems like a good upper level. (note I'd also change the text for the monthly amounts to something better)

**Comment by scd** [comment]
The 'Once off' and 'Monthy supporter' radio buttons are perfect and in line with my expectations. In this case we are compromising with some hidden fields but this is what I expected. I am very happy with the UX you described above.

I think we have now discussed the fully enough and become exited about this piece of work that I should fully implement all dimensions as described.

Initially I thought the minimum monthly should be $10 otherwise to big a proportion goes to stripe. If it was $5 then we are much better off getting them to do a once off $60 transaction. I was thinking we have $10, $20, $50 + other.

If somebody is doing a monthly payment then they are probably working full time, in which case $10 - $50 is not unreasonable if they are blown away by thecrag.

Uni students I would expect to do the minimum once off.

Sorry I did not mean to say that you did not know how stripe worked. Your comment about people filling in details every month threw me. But in hindsight I see that you were just making a point.

**Comment by brendanheywood** [comment]
all good :)

In my mind someone who gives $60 a year in monthly instalments is better than someone who gives $60 in a single hit. The former I reckon is much more likely to not churn at the end of the year and just keep going next year. The fraction that goes to stripe is still way way lower than say Apple stores or Google plays cut at 30% so I don't think should be a major part of the decision. It's $58.65 vs $55.35 and if it means they are more likely to stick around for a second year then it's $58.65 vs $110.70. I think the major determiner should be what people are most likely to consider paying and we should base that off the history of current donations and refine it over time as we get more. That said I'm still happy to go $10, $20 etc for monthly as long as it is for the right reasons. The jump from $20 to $50 seem big through, $240 a year to $600 a year, I'd probably go $10, $20, $30 Other (I regularly donate to 5 charities and all are around the $30 a month mark, seems to be a common sweet spot). Again we'll refine this over time with real feedback

**Comment by scd** [comment]
I have now create a user as a stripe customer when we make charge. We can then make subsequent charges using the last credit card without the user typing out the whole card number.

I have some test modes that makes it easier to review various user states. These states overwrites some template data.

dev.thecrag.com/donate?mimic=never-supporter
dev.thecrag.com/donate?mimic=supporter
dev.thecrag.com/donate?mimic=renew-supporter
dev.thecrag.com/donate?mimic=reactivate-supporter

The form looks a lot nicer using bootstrap forms, but the cvc and credit card input field sizes are overriden.

Things still to do:
 - [ ] redirect to contributions page - currently easier to test without this
 - [ ] decide whether or not to allow a comment. Nice idea, but realistically we are not going to read it, unless it creates a pm to us.
 - [ ] decide whether to have a blank month field
 - [ ] test on mobile - hopefully the bootstrap form just works, otherwise we may have to make the fields stacking
 - [ ] monthly payments $10, $20, $30, other (create plans on the fly)
 - [ ] monthly payments webhooks so the system knows when a payment has been made
 - [ ] how to test webhooks for monthly payments
 - [ ] change monthly payment plan
 - [ ] unsubscribe to monthly payments
 - [ ] test live mode in dev
 - [ ] verify email receipt in live mode after payment

**Comment by brendanheywood** [comment]
> Nice idea, but realistically we are not going to read it, unless it creates a pm to us.

Why not just send en email to support every time a new supporter signs up? I'd love to read these, and I'd certainly love to know when it happens. In fact why not just make this the normal process (which integrates the manual thank you email that you send):

1) someone becomes a supporter, with optional message
2) system sends an email to support with the optional message, with a reply-to of the persons email, using either the receipt email or their account email
3) we then treat it exactly like a priority support email, we all see it, and can all reply if needed depending on the comment details eg "here is $100 I'd really like if it if you guys fixed XYZ"
4) after this if they respond it becomes an email thread no different to if they had directly emailed us the first time

Priority support is one of the few tangible benefits we can offer right now for supporters

> test on mobile - hopefully the bootstrap form just works, otherwise we may have to make the fields stacking

I've fixed up the responsive layout

> how to test webhooks for monthly payments

Stripe has a test webhook gui and you can emulate them at any time. Also its important to treat the webhooks as notifications only, ping their api to confirm details and handle duplicate hooks properly.

* [ ] I also find the 'other' behaviour a bit weird when it jumps around after I type something in. I'd prefer instead for it to be a radio button itself and if you focus on the text it selects the radio button as a convenience. Nothing every moves or changes

**Comment by scd** [comment]
> Why not just send en email to support every time a new supporter signs up

I will buy into that as a requirement. It is actually a nice value proposition and may result in somebody donating a large amount. This comment will go onto the email. There is no place to store in in our system, but we could store it in meta data associated with the payment.

Also I am in two minds whether I should be pm'ing these guys after they donate. My current process is to PM only after their first donation. I am not so keen about automating this because I like to look at their account and maybe I go to their blog if it is listed on their profile. Sometimes this results in a very different flavor. If we are doing the email, then anybody who is interested in a conversation will add their comments. So I think I am going to stop PMing.

>     *  I also find the 'other' behaviour a bit weird when it jumps around after I type something in. I'd prefer instead for it to be a radio button itself and if you focus on the text it selects the radio button as a convenience. Nothing every moves or changes

Better.

**Comment by brendanheywood** [comment]
I just fixed up the other label and focus behaviour.

> Also I am in two minds whether I should be pm'ing these guys after they donate. My current process is to PM only after their first donation. I am not so keen about automating this because I like to look at their account and maybe I go to their blog if it is listed on their profile. Sometimes this results in a very different flavor. If we are doing the email, then anybody who is interested in a conversation will add their comments. So I think I am going to stop PMing.

To me you are still PM'ing, it's just how you go about it is better and more visible to the rest of us, and it's easier for us to jump in if you are away etc. To be clear I only meant this email would directly go to support, not them, and then we would manually and personally reply to it (eg after we've looked at their profile or whatever). This would be separate to the receipt email which I believe stripe sends for us? If stripe doesn't send this and we have to do this ourselves then perhaps we could merge these into an email which gets sent to them with their first receipt, and cc's us, and we could then reply with a more personal follow up.

**Comment by scd** [comment]
btw I had to merge some changes you made with some of my changes. Not a problem, but I am only mentioning it because wrapping the label around both the radio button and other input causes problems at least on my browser. I am unable to type into the text field. I am sure this is overridable, but the way I solved it was to use an inline radio and not wrap them.

**Comment by brendanheywood** [comment]
which browser? maybe wrap the label around just the inline help part and not the input itself

**Comment by scd** [comment]
Awsome, I have got the beginnings of monthly subscriptions in.

 - [x] UX for new subscripition
 - [x] create a plan on-the-fly based on the person's subscription, if one does not already exist
 - [x] subscribe a person to a plan
 - [x] if they are already subscribed to a plan then change plan with no prorata adjustments (ie just get billed for the larger/smaller amount next time).
 - [ ] update email address in stripe and get email address from stripe customer record
 - [ ] cancel a subscription
 - [ ] clear indication of what subscription a person is on if already subscribed + on contributions page
 - [ ] webhooks

**Comment by scd** [comment]
Getting closer to a minimal viable product.

I just discovered that once somebody subscribes to a plan for a particular currency they cannot change currencies. I think the way around this is to create a new customer. The code now caters for this currency restriction, however changing currencies after you subscribe is an edge case so I think I will leave it for a while. Maybe removing an existing customer linkage is an admin function.

**Comment by scd** [comment]
@brendanheywood a thought just came to me while we are going to town with this stripe integration. Should we also add google events to stripe payments?

I think we need to know, errors, card payment, card reuse, subscription, subscription charge, charge failure. Maybe we just map all the stripe webhook events to ga. 

We could fake a url with stripe event name and account id. Actually the url would be semi real because it is a real webhook url /apps/stripe. The faked bit could be /apps/stripe/charge.succeeded

We also probably need a separate event which tracks the amount charged.

**Comment by scd** [comment]
@brendanheywood could you please pull from repo. I have made the webhooks implementation but need to test it on dev.

**Comment by scd** [comment]
ga integrated. I have decided to map all stripe events sent out by webhooks to google analytics using a fake uri extension on the webhook api. I think this is cool. I have also added an '/error/error-message' to the end of the fake uri if there is an error.

Stripe charges are also mapped to ga transactions. I am not sure what to do with this at GA end just yet, but I am sure there will be some brilliancy.

**Comment by brendanheywood** [comment]
sorry kinda been on a tassie holiday :)

pulled and deployed anyway. Are you doing this ga stuff client side or server side? Either way need to ensure the userid stuff is set properly. As a general rule if it is important business data then it should be in our app, and nothing sensitive in GA. we shouldn't be using ga as a data store. The primary benefit I can see for GA having this is we can set up proper funnel's to see how people / traffic flows to the donate page.

With the GA transactions did you also add the extra tracking code and also turn this feature on in the GA admin settings in both ga accounts?

**Comment by scd** [comment]
Holiday's are good. 

I was doing some server side stuff, but based on your comments I decided to do some extra client side stuff as well. I have just pushed to repo.

So there are three places where I have added ga:
 - [x] client side for success/fail of charging (ie once they are in the donate page) - this has the uid
 - [x] server side on stripe webhooks to see what webhooks and errors are coming in, no uid as this is a little cumbersome to get.
 - [x] server side for the actual transaction, which may be the once-off charge or a monthly subscription charge (in which case we log a transaction every month). THis has uid.

I don't really know what you mean by using ga as a data store. I do want to use ga for tracking more stuff other than just page views. In particular I am keen to get system events into ga so we can track (and compare) performance of specific events over time. In that vein I want to track money as well so we know how our dollars compare to traffic.

I have now enabled standard ecommerce for dev and main ga properties

**Comment by brendanheywood** [comment]
> I don't really know what you mean by using ga as a data store

All I mean by this is that if some particular piece of data is mission critical, we should be storing it locally in mysql and not externally (or in both is fine). ie if google shut down GA tomorrow, it would be annoying as hell to *us*, but the application would keep on working. I still haven't had a good play with the whole thing so I'll do that and have some feedback over the weekend

**Comment by scd** [comment]
> we should be storing it locally in mysql and not externally (or in both is fine). 

Yup. Transactions stored in both, other stuff not mission critical.

Also I have been thinking deeper about emailing. I am leaning toward not using stripe's automated emails but rather using our own.
 * use our own standard email template is better for our brand
 * we can bcc each new once-off payment and sponsorship to our support email which fulfills another requirement of notifying us
 * I can make it a GST receipt in the case of payments to an AUD currency, which is really an accounting requirement and a pain for me to do separately for the odd business that can make something of a GST receipt.

**Comment by brendanheywood** [comment]
Yeah I like the custom email's too.

**Comment by scd** [comment]
@brendanheywood can you pull again from repo before you start testing. I need to test emails which I cannot properly do on my dev.

**Comment by brendanheywood** [comment]
done

**Comment by scd** [comment]
FYI, I am still doing lot's of testing. There are plenty of bugs I am working through. Notes to self:

 - [ ] user message needs to be included in email
 - [ ] cc support@thecrag.com
 - [ ] cancel email
 - [ ] change subscription email

**Comment by brendanheywood** [comment]
what the hell? I posted a comment yesterday with a big list (maybe 20+ items) of feedback and tasks for stripe, but it's not here. I'm pretty sure I sent it but maybe I'm going mad. Did you get an email with it? It may have been accidentally deleted so if you have the email please copy it back in here. Doh!

* [ ] Also not critical for right now for first release but we'd also want a card expiry email, eg

![image](https://cloud.githubusercontent.com/assets/187449/6198602/6887a6fe-b461-11e4-84b0-b5a47831e117.png)

**Comment by scd** [comment]
never saw it. I think you are going mad :)

**Comment by brendanheywood** [comment]
Redone off the top of my head, more or less from the top of the form to the bottom of the form:

* [ ] yellow box at top, text only makes sense for one-off donations, it doesn't read right if you regularly donate. "You are up-to-date with your financial support until Thu 18th Jun 2015" - NO I am a financial supporter forever until I cancel.
* [ ] if you are on a plan, the yellow box and the 'active supporter plan' section are more or less the same thing so ditch one or the other. I'd keep the yellow box but flesh it out with the $ amount and the date when you will next be charged.
* [ ] other - should validate onchange rather than onsubmit
* [ ] 'tell us what you feel' the placeholder texts feels a bit snarky to me, maybe something "Extra comments: Tell us what you love or hate, or if there is something in particular you'd like us to focus on'
* [ ] receipt email - I've had a look at a few other places I store my CC with, and most don't ask for email they just use my account email without asking. Given that we only only allow logged in accounts I'm wondering if there is any point to this field.
* [ ] name on card - against if this isn't really used by stripe, and we don't care, why not just remove this field as well?
* [ ] 'Stripe secure payment' stripe isn't a house hold name like say paypal. Most people who use stripe don't mention them at all and those words are a little confusing. Maybe "Secure payments using stripe.com" or just remove it completely.
* [ ] in the payment history can we show whether it was a one-off or a regular payment. we may as well show other stuff in this column like receipt number

* [ ] after submission I find the 10 second delay and reload weird. I reckon either go to a completely new thanks you page, or reload this page so it shows the new layout for current supporters

If a current supporter:
* [ ] if you click cancel, and then click once-off the button at the buttom still says cancel
* [ ] if you are a current supporter the button should say 'update subscription' instead of subscribe
* [ ] it should be super clear when your next payment will be and how much it will be. should be right at the top, probably in that yellow box
* [ ] if you are a regular supporter, we still allow once off payments, that's cool but might be a fringe case. should make it clear this will be imediate and won't affect your next refular payment. I'd be happy to remove the once-off if they are a regular
* [ ] CRITICAL: on the opposite case if you have previously made a once-off, and then start a new monthly payment I would expect this to happen right away, the first payment would be right this second and then another one in a month. This doesn't seem to be happening. I have no idea when my next payment will be, it could be in a year, or a month, or something else. This use case I think is really important as most of the people who will first sign up as regulars will have also donated in the last year. This needs to be smooth.
* [ ] if you have a credit card stored there is no way to remove it. Cancelling a supporter plan doesn't remove it but probably shouldn't. I think there should just be a 'Delete this card' button next to the card last 4 digits

In fact if you are a current supporter I think it might be best to not try and repurpose this form at all, and just have a completely different simpler layout. Almost everything on that form is redundant for current supported. It should just have an update plan radio button, and a cancel plan button.

Mockup:

![image](https://cloud.githubusercontent.com/assets/187449/6198716/45f0a30a-b468-11e4-9402-b84020d3ccf1.png)

**Comment by brendanheywood** [comment]
In that mockup above, if you chose cancel the payment details could shrink away and be replaced with a comments box asking why they have cancelled.

**Comment by brendanheywood** [comment]
Some little tweaks:

* [ ] theCrag.com is a registered business in Australia (ACN: 090 494 698).

Can we just move this into the email footer and make it universal, and not part of this main email body text?

* [ ] Also reciept should be receipt

* [ ] Also it might be a fringe issue but should we also include the username of the donor account in the details? Just thinking for emails like this where the email is shared, of if you have multiple accounts.

* [ ] should have a link back to donate page like the other emails

**Comment by scd** [comment]
There are four different support emails that we send out.
 - [x] Once off payment charge, which happens live on submission.
 - [x] Starting a subscription, first charge which happens about 10 seconds after the subscription has been created via stripe webhooks
 - [x] Monthly subscription charges (other than first charge), which happens via stripe webhooks.
 - [x] Cancel a subscription, which happens live on submission
 - [x] Update a subscription, which happens live on submission

After we have demonstrated that people are actually using the plan (ie not version one) this should be extended to:
 - [ ] Expired credit card notification when on a subscription
 - [ ] Suspend a plan
 - [ ] Resume a plan (note that when you resume a suspend plan it does not charge until the next billing date, which is slightly different from cancelling and starting a new plan which will charge immediately).

**Comment by scd** [comment]
Do we want to allow partial dollars? Let's just go with whole dollar amounts.

**Comment by scd** [comment]
I have made a few fixes, please pull from repo. The cc stuff will not work on your dev until I fix something, just let me know if you want to see this working on your dev.

Also I am not going to implement everything in this version because I simply don't believe that many people will go onto the monthly payment plan. If we get 10 people on this plan then I will change my tune.

> receipt email - I've had a look at a few other places I store my CC with, and most don't ask for email they just use my account email without asking. Given that we only only allow logged in accounts I'm wondering if there is any point to this field

We don't force email, nor do we get people to validate their email. We probably should do this for various reasons, but until we do I think we need the email field.

 > in the payment history can we show whether it was a one-off or a regular payment. we may as well show other stuff in this column like receipt number

There is no place to store payment type in the table as it is currently defined. As soon as I feel that it is important I will make the change.

> after submission I find the 10 second delay and reload weird. I reckon either go to a completely new thanks you page, or reload this page so it shows the new layout for current supporters

This will be the last thing I change because it makes it easier for testing to have it go to here. A while ago you suggested going to the settings/contributions page. There is an issue reloading the page when subscribing to plans - see below.

> CRITICAL: on the opposite case if you have previously made a once-off, and then start a new monthly payment I would expect this to happen right away, the first payment would be right this second and then another one in a month. This doesn't seem to be happening. I have no idea when my next payment will be, it could be in a year, or a month, or something else. This use case I think is really important as most of the people who will first sign up as regulars will have also donated in the last year. This needs to be smooth

Although stripe makes the charge immediately the actual charge id is not returned with the subscription. The charge comes through webhooks about 5-10 seconds later. The webhook triggers recording the payment and the email. It was probably not working for you because there were bugs in the repo version.
So far I have not found an easy way to find the charge ID synchronously when creating a subscription (note that our payment transaction records link to stripes charge ids) - which is my ideal solution. See api:

https://stripe.com/docs/api/curl#subscriptions

So I am using asynchronous webhooks. The only downside to all this is an asynchronous delay of about 10 seconds.

Once you have pulled the most recent repo there is still some more testing I need to do on our dev.thecrag.com.

**Comment by brendanheywood** [comment]
pulled

>We don't force email, nor do we get people to validate their email. We probably should do this for various reasons, but until we do I think we need the email field.

Ok then

> Although stripe makes the charge immediately the actual charge id is not returned with the subscription. The charge comes through webhooks about 5-10 seconds later.

It doesn't need to be synchronous. We could just go to the next page as soon as the subscription is good and in the table show something like:

![image](https://cloud.githubusercontent.com/assets/187449/6207883/024035d0-b602-11e4-802e-fe131e8c0be3.png)

In fact does stripe even guarantee any sort of times around this? I'd imagine under some conditions the webhook could turn up much later

**Comment by scd** [comment]
@brendanheywood thanks for all your feedback on this. I think we are very close, could you please pull from repo and do a final review. Once you and I are happy I will get Nicky and Campbell to review.

Please note that I plan to leave following tasks until a later version. Please feel free to argue otherwise.

 - [ ] No Pending status in the historical payments table. Instead this is clearly marked in the yellow box at the top of the file. I am leaving this until the requirements for the account payment table have stabilized. In this particular instance the requirement is for a known payment which is not yet confirmed by the payment gateway.
 - [ ] Not adding a payment type to the historical payments table (see above).
 - [ ] Not storing the user comment anywhere except for the email. We don't have any table for this sort of information and it is not critical to store, but is critical to read. If we do end up wanting to store this then maybe I will do something similar to wordpress and build an account meta table where we can store anything. We have got something similar already, but it is storing stuff that is then available in the account template data. In other words more thought.
 - [ ] Default month is set to current month. I am looking at https://secure.karelia.com/buy_now/ as the benchmark (the example brendan gave early in this issue). There is a 1/12 correlation for month and about 1/3 correlation for year. I figure the argument is not about the correlation, but the form first impressions (it looks neater filled in).
 - [ ] Pay now button not disabled when fields are not filled in properly. Besides my code not being sympathic to this requirement I am not so keen on this kind of functionality, for a couple of reasons. Firstly the fail safe is that people cannot submit if the button is disabled, secondly we are throwing validation errors into GA, so it is more useful to know where people are actually going wrong. Thirdly if somebody misses a field then it can be very frustrating looking at a disabled button, whereas validation error text when they press the button is more user friendly.
 - [ ] I think there is an argument for validating the 'other' amount field on entry, as I am not doing any live validation I feel as though I don't want to do a special for one particular field. While it is an improvement in the user experience it is fairly minimal and would not result in any more payments IMO.
 - [ ] Suspend/resume a plan. I just want to wait until we have got enough people on a plan to justify me spending my time on this. It is not an issue for convincing people to go on a plan in the first place.
 - [ ] Expiring credit card notification. Ditto.
 - [ ] Instead of 'Stored card' I am using the wording 'Last card'. We do not store card numbers so we have to be careful on the wording here. If we want to use the phrase 'Stored card' then we need a footnote.

**Comment by brendanheywood** [comment]
done. Everything is cool, tweaked some whitespace in the form

I'm really happy with it in it's current state and it is definitely ready to ship.

**Comment by nicHoch** [comment]
**Nicky 's Test Report**

- looks good in general I was able to do two donations: onetime 60€ and monthly 10€. each with a separate CC
- some play around with false numbers dates .. working fine
- after "coming back" from the first donation the header massage changed to "thanks ..." and in the donation history the transaction was visible - very good
- after "coming back" from the second donation the new transaction was not shown in the history right away - after reloading the page it was added to the history
- the mail was correctly send to my account mail but my Thunderbird does not like the content: it as labeled as SCAM because the link label and the real link address does not match. Every link is redirected over mandrillapp.com. This link: 

```
You may also find your account profile page here:
https://www.thecrag.com/nickyhochmuth
```
is not working at all - the other 3 (2 in footer) find there target.

![screenshot 356](https://cloud.githubusercontent.com/assets/2226714/6226399/0fe55d94-b695-11e4-9d42-d934b90bed9e.png)

- General thought about CC service. In Europe the people are not that trained and confident with spreading CC informations around. I guess it is important to know for an anxious user who exactly is storing the CC data and who takes care for the transaction. A Paypall button is implicit pretty clear in that. In the current version it is not visible for the user what service we use.
	* no mention of stripe at all
	* does Stripe the Transaction?
	* is theCrag storing any CC data on its own servers?
	* could I delete my stored CC Data 

**But bottom line: I like the stripe integrations it offers good opportunities for more flexible reimbursement strategies.   **

**Comment by scd** [comment]
Notes to fix:

 - [ ] Footnote at the bottom about the payment gateway and not storing CC.
 - [ ] Fix the link to profile page.

Also not sure what to do about the email issues. @nicHoch can you request your password be reset on the main server - this should trigger an email. Please tell us whether that say it is a scam.

**Comment by nicHoch** [comment]
password reset done: **also marked as SCAM**

can we not provide an own click-tracking and service? 
Is GA not capable to track from where the user comes from?

**Comment by brendanheywood** [comment]
We already have the ga params in use in some emails. I'd be happy to drop the mandril link stuff if it turns out that it's a factor in why it's being marked as a scam. I find it slows the links down a bit too much with the extra redirect as well. @nicHoch for a proper test do you also see that warning if you reset your password in production? this may be limited to the dev box

**Comment by nicHoch** [comment]
both dev and prod SCAM.
This is a very common (but simple) attack pattern. This SCAM warning will not show up if link label and link address are the very same or the link label is not an url. But to add the entire mandrill stuff to the label is not an option. on other workourd would be:

... can be found <a href="">here</a>

**Comment by scd** [comment]
where is here - the link above is a reference to this issue

**Comment by nicHoch** [comment]
with here i mean a pattern:

```
//this is NOT filtered as SCAM
<a href="http:\\some.url">here</a>  

//this is filtered as SCAM
<a href="http:\\some.url">http:\\someother.url</a>  
```

**Comment by scd** [comment]
This is a good pickup Nicky. It is important for us to expose the actual urls and mandril click tracking is not important.

I have turned off mandril click tracking in Mandril. Actually it was an oversight that it was ever in there.

@nicHoch can you please retest. I have also fixed the other issues you raised.

**Comment by nicHoch** [comment]
- now mail is not marked as SCAM.
- Other currencies are gone - why
- I was not able to donate less than 30$! To be frank thats sounds a bit snobby to me. Do we not like and encourage also small donations.

**Comment by scd** [comment]
> Other currencies are gone - why

This is a stripe constraint. Once a stripe customer signs up to a plan with a particular currency they cannot change currencies. I have deleted you as a stripe customer so you should be able to test again. The only way around this is to drop theCrag account/Stripe customer association and create a new customer. I don't imagine many people will want to swap currencies on their payment plans, but if it becomes a problem then we need another function. But for now I am happy to leave it as is.

**Comment by scd** [comment]
to do:

 - [ ] make it clear what the minimum other amount is.

**Comment by scd** [comment]
I have made the minimum other amount explicit as I think it would be very annoying for somebody to type in, say $20, then get rejected. If the minimum amount is shown then somebody will know before they make the decision to donate or not. This will go someway to alleviate the snobby issue without us actually changing our minimum amount.

**Comment by scd** [comment]
I am closing this issue now. After the release I will have to do some testing immediately. I will make the minimum once off payment as a validation test. I will assume that the monthly payments will just work if the once off does. I have noted this in the milestone release notes.
